### PR TITLE
fix the ratelimiter

### LINKED ISSUE #1763: Rate Limits are being hit, Bot is being blocked, IP Banned
Thank you very much for your solid work on DSharpPlus, It has been magical for me to work with.

### Background: 
The bot acts as a Killfeed Bot for a video game, every action in the game posts a message to a designated discord channel for their players to see. The bot is public and is currently connected to approx ~400 gameservers. Some "battlefield" type gameservers have ~2-3 kills per second. The bot is self hosted at my home, with a static IP. As you can imagine, its in the nature of the beast to exceed rate limits by posting each kill event into a discord channel.

### Describe the bug
I was under the impression that this library would queue messages being posted until a future timeframe where the rate limit would not be exceeded, this does not seem to be the case. For the last two days running, the bot has caused my IP to be temporarily banned during peak times.

This is the code used to post messages to discord.

```cs
try
{
    if (await DiscordClient.GetShard(gs.GuildId).GetChannelAsync(gs.KillfeedPvEChannel.Value) is DiscordChannel chnl)
    {
        var eb = new DiscordEmbedBuilder(); // other logic to build the embed hidden

        await chnl.SendMessageAsync(x =>
        {
            x.AddEmbed(eb);
            x.WithAllowedMentions(Mentions.None);
        });
    }
}
catch (UnauthorizedException e)
{
    logger.LogWarning(e, $"Removing PvE Channel, Unauthorized, for GuildID: {gs.GuildId}, GameserverID: {gs.Id}, Channel: {gs.KillfeedPvEChannel.Value}");
    // remove the PvE Discord channel from the DB if we cannot send messages in there to stop future posts
    gs.KillfeedPvEChannel = null;
}
catch (NotFoundException e)
{
    logger.LogWarning(e, $"Removing PvE Channel, NotFound, for GuildID: {gs.GuildId}, GameserverID: {gs.Id}, Channel: {gs.KillfeedPvEChannel.Value}");
    // also remove the PvE Discord channel if it has been deleted
    gs.KillfeedPvEChannel = null;
}
catch
{
    // catch all other errors and continue
}

await db.PvEEvents.AddAsync(new Entities.PvEEvent()
{
    Entity = msg.PvEFeedType,
    GameserverId = gs.Id,
    PlayerId = player.Id,
    TimeStamp = DateTimeOffset.UtcNow.ToUnixTimeSeconds()
});

await db.SaveChangesAsync();
```
### To Reproduce**
Ummm.... be a super spammy bot¿? and post a tonne of messages to lot of different guilds and watch being temp IP banned.

I'm happy to screen share the code, but its not in a public repo.

### Expected behavior
That the library would "queue" messages and not send them if they were to hit the rate limit.

### Screenshots
```
[2024-02-13 17:28:13 +11:00] [114 /RatelimitHit] [Error] Ratelimit hit, requeueing request to https://discord.com/api/v10/channels/**************30558
[2024-02-13 17:28:13 +11:00] [114 /RatelimitHit] [Error] Ratelimit hit, requeueing request to https://discord.com/api/v10/channels/**************79742
[2024-02-13 17:28:14 +11:00] [114 /RatelimitHit] [Error] Ratelimit hit, requeueing request to https://discord.com/api/v10/channels/**************743946
[2024-02-13 17:28:14 +11:00] [114 /RatelimitHit] [Error] Ratelimit hit, requeueing request to https://discord.com/api/v10/channels/**************99807
[2024-02-13 17:28:14 +11:00] [114 /RatelimitHit] [Error] Ratelimit hit, requeueing request to https://discord.com/api/v10/channels/**************086726
[2024-02-13 17:28:14 +11:00] [114 /RatelimitHit] [Error] Ratelimit hit, requeueing request to https://discord.com/api/v10/channels/**************425396
[2024-02-13 17:28:14 +11:00] [114 /RatelimitHit] [Error] Ratelimit hit, requeueing request to https://discord.com/api/v10/channels/**************085466
[2024-02-13 17:28:16 +11:00] [114 /RatelimitHit] [Error] Ratelimit hit, requeueing request to https://discord.com/api/v10/channels/**************53536/messages
[2024-02-13 17:28:16 +11:00] [114 /RatelimitHit] [Error] Ratelimit hit, requeueing request to https://discord.com/api/v10/channels/**************53536/messages
[2024-02-13 17:28:16 +11:00] [114 /RatelimitHit] [Error] Ratelimit hit, requeueing request to https://discord.com/api/v10/channels/**************53536/messages
```
image

### PR DESCRIPTION
**Pull request by akiraveliara**
this has grown way out of the original bounds of the PR, but here we are.

this fixes the ratelimiter under high stress, solving the main bug we struggled with as well as a number of smaller issues that may have contributed edge cases.

fixes #1763

### CODE REVIEW

File: DSharpPlus/Net/Rest/RateLimitStrategy.cs [hunk 1]
```diff
@@ -1,35 +1,37 @@
...
+
+    private readonly ILogger logger;
+    private readonly int waitingForHashMilliseconds;
+
+    public RateLimitStrategy(ILogger logger, int waitingForHashMilliseconds = 200)
+    {
+        this.logger = logger;
+        this.waitingForHashMilliseconds = waitingForHashMilliseconds;
+
+        this.ratelimitCleanerTask = CleanAsync();
```
Comment (Plerx2493) [inline DSharpPlus/Net/Rest/RateLimitStrategy.cs:33]: Do we need to dispose this task somehow if this gets disposed

**Review by Plerx2493 — COMMENTED** [review]
LGTM, some nits

File: DSharpPlus/Net/Rest/RateLimitStrategy.cs [hunk 2]
```diff
@@ -55,137 +59,166 @@ TState state
...
+    private Outcome<HttpResponseMessage> SynthesizeInternalResponse(string route, DateTime retry, string scope)
+    {
+        HttpResponseMessage synthesizedResponse = new(HttpStatusCode.TooManyRequests);
+
+        synthesizedResponse.Headers.RetryAfter = new RetryConditionHeaderValue
+        (
+            retry + TimeSpan.FromMilliseconds(Random.Shared.NextInt64(50))
+        );
+
+        synthesizedResponse.Headers.Add("DSharpPlus-Internal-Response", scope);
```
[the remark this answers is not in the document]
Comment (akiraveliara) [reply DSharpPlus/Net/Rest/RateLimitStrategy.cs:162]: it is

File: DSharpPlus/Net/Rest/RateLimitStrategy.cs (hunk 1 above)
Comment (akiraveliara) [reply DSharpPlus/Net/Rest/RateLimitStrategy.cs:33]: i suppose we should, yes

**Review by Plerx2493 — APPROVED** [review]

**Pull request merged**